src/docs/markdown/extending-caddy/placeholders.md

```diff
@@ -0,0 +1,140 @@
+---
+title: "Placeholder Support"
+---
+
+# Placeholders
+
+In Caddy, placeholders are processed by each individual plugin themselves. They are not parsed at config time, but instead preserved and processed at runtime.
+
+This means that if you wish for your plugin to support placeholders, you must explicitly add support for them.
+
+If you are not yet familiar with placeholders, start by reading [here](/docs/conventions#placeholders)!
+
+## Placeholder Internals
+
+Internally, placeholders are simply a string in format `{foo.bar}` used as valid configuration values, which is later parsed at runtime.
+
+Placeholders-like strings which start with a dollar sign (`{$FOO}`), are evaulated at Caddyfile parse time, and do not need to be dealt with by your plugin. This is because these are not placeholders, but Caddyfile-specific [environmental variable substitution](/docs/caddyfile/concepts#environment-variables), they just happen to share the `{}` syntax.
+
+It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`
+
+As an example, see the following caddyfile:
+```caddyfile
+:8080 {
+  respond {$HOST} 200
+}
+
+:8081 {
+  respond {env.HOST} 200
+}
+```
+
+When you adapt this Caddyfile with `HOST=example caddy adapt` you will get
+```json
+{
+  "apps": {
+    "http": {
+      "servers": {
+        "srv0": {
+          "listen": [
+            ":8080"
+          ],
+          "routes": [
+            {
+              "handle": [
+                {
+                  "body": "example",
+                  "handler": "static_response",
+                  "status_code": 200
+                }
+              ]
+            }
+          ]
+        },
+        "srv1": {
+          "listen": [
+            ":8081"
+          ],
+          "routes": [
+            {
+              "handle": [
+                {
+                  "body": "{env.HOST}",
+                  "handler": "static_response",
+                  "status_code": 200
+                }
+              ]
+            }
+          ]
+        }
+      }
+    }
+  }
+}
+```
+
+Importantly, look at the `"body"` field in both `srv0` and `srv1`.
+
+Since `srv0` used `{$HOST}`, the special environmental variable replacement with `$`, the value became `example`, as it was processed during Caddyfile parse time.
+
+Since `srv1` used `{env.HOST}`, a normal placeholder, it was parsed as its own raw string value, `{env.HOST}`
+
+Some users may immediately notice that this means it is impossible to use the `{$ENV}` syntax in a JSON config. The solution to this is to process such placeholders at Provision time, which is covered below.
+
+
+## How to use placeholders in your plugin
+
+#### Parse the raw placeholder value in your unmarshaler
```

Comment:
I don't really think this section adds any value / says anything interesting. It can probably be removed...

Reply:
hmmm, my thought is that some people may be confused thinking they might need to use a special function to parse a placeholder.

if you don't think thats a problem, we can gut it

Reply:
What if we remove the section but put what you're conveying in the parent section ("Implementing support for placeholders" as a heading I suggest):

> Placeholders are evaluated at runtime (not at Caddyfile parsing or config loading). When you are unmarshaling a Caddyfile, do not evaluate placeholders at that time. Placeholders may be evaluated during `Provision()` or during your module's execution.